### ISSUE TITLE
Cannot insert explicit value for identity column when ValueGenerationStrategy is set to IdentityColumn

### ISSUE THREAD

**Issue by rbluethl** [dotnet/efcore#15220]
Hey there,

I have a table with ValueGenerationStrategy set to `SqlServerValueGenerationStrategy.IdentityColumn`. When adding a new item to the DataContext, the value is normally generated automatically. However, we have implemented a mechanism that a database action is automatically retried when an attempt fails. It's basically just a queue of actions which cyclically processes all failed actions.

```c#
ExecutionStrategy.Execute(() =>
{
  try
  {
    using (var dataContext = DataContextFactory.GetContext())
    {
      dataContext.AlarmLogs.Add(alarmLog);
      dataContext.SaveChanges();
    }
  }
  catch (Exception ex)
  {
    throw new DatabaseException("Could not add alarm log.", ex);
  }
});
```

When something fails within the `Execute` block, we're adding the action to a queue and try it again. So it apparently happens, that when an action is retried for whatever reason, the `Id` property of the `AlarmLog` class is set to `Int64.MinValue` at some point. I can confirm when I have a look in the debugger. But since the Id is automatically generated, the `SaveChanges()` call then fails with the following exception. I already read somewhere that this is expected behavior and that the ID is generated during SaveChanges, however it doesn't seem to work reliably. When I MANUALLY set the Id to zero before saving like this:

```c#
// Always set to zero manually
alarmLog.Id = 0;
dataContext.AlarmLogs.Add(alarmLog);
dataContext.SaveChanges();
```

... then everything works just fine. Am I missing some annotation on my table? I kind of doubt it, because as I already mentioned, normally it works as expected. Is there a problem when retrying the entire action in a different thread? I always get a new `DataContext` as you can see in the code above, so I guess that shouldn't be a problem.

Am I doing something wrong?
Thank you!

Ronald

```
Exception message: `Cannot insert explicit value for identity column in table 'alarm_log' when IDENTITY_INSERT is set to OFF.`
Stack trace: ```at Microsoft.EntityFrameworkCore.Update.ReaderModificationCommandBatch.Execute(IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(DbContext _, ValueTuple`2 parameters)
   at Microsoft.EntityFrameworkCore.SqlServer.Storage.Internal.SqlServerExecutionStrategy.Execute[TState,TResult](TState state, Func`3 operation, Func`3 verifySucceeded)
   at Microsoft.EntityFrameworkCore.Update.Internal.BatchExecutor.Execute(IEnumerable`1 commandBatches, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Storage.RelationalDatabase.SaveChanges(IReadOnlyList`1 entries)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(IReadOnlyList`1 entriesToSave)
   at Microsoft.EntityFrameworkCore.ChangeTracking.Internal.StateManager.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges(Boolean acceptAllChangesOnSuccess)
   at Microsoft.EntityFrameworkCore.DbContext.SaveChanges()
   at DataAccess.Repositories.Alarms.AlarmRepository.<>c__DisplayClass4_0.<AddAlarmLog>b__0() in ...:line 86```
```

### Further technical details
EF Core version: 2.2.3
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10 Pro
IDE: Visual Studio 2017 15.9.10

**Comment by ajcvickers** [comment]
The root cause of this is the same as #12378. However, @rbluethl, you may want to look into EF [connection resiliency](https://docs.microsoft.com/en-us/ef/core/miscellaneous/connection-resiliency) for these kinds of retries.

**Comment by rbluethl** [comment]
Hey @ajcvickers, thanks for the quick response. I understand -- however, is there a recommended workaround for now? Is it OK to set the Id to 0 before saving? Will this fix already be included in the latest 3.0.0 preview version? Is there already an ETA for 3.0.0?

Thanks for pointing out connection resiliency. I did indeed try to use it, however, I found that it doesn't exactly suit my needs. From what I can tell is that the calls are blocked and retried multiple times until they ultimately fail. What I need to do is try an execution and, in case of an exception, add them to a separate queue to prevent the thread from blocking until they succeed. Is this possible? If so, I would be happily using it.
